### DISCUSSION TITLE
Id type support on a per-collection basis

### DISCUSSION THREAD

**Discussion by JPrisk** [payloadcms/payload#5110] [Feature Requests & Ideas]
🙌 great to see the option of `uuid`'s get added

following on the back of #3864, it'd be great if users could configure the `idType` on a per-collection basis

- it would help with incremental adoption of this new feature
- serial `id`'s and `uuid`'s have slightly different use cases, each offering benefits - it'd be nice to get the best of both worlds

this request might introduce some complexity given `@payloadcms/db-mongodb` doesn't support the new `idType` arg, therefore it would need to be conditionally supported in the collection config or coincide with support for `idType` for MongoDB

opening for consideration 🙏🏼

**Comment by DanRibbens** [comment]
This could be done now with some creativity.

Leave the id type alone so the serial IDs are the default.

When you need a UUID id type add your own id field:
`name: 'id', type: 'text', admin:{ hidden: true }`

Then with a before change hook on the field you can generate return and return a UUID. It is going to be a varchar field in the db, but functionally it is the same.

**Reply by JPrisk** [reply]
thanks for the reply @DanRibbens, I had a thought to do something similar - guess the api would still use `id` as a primary key which is what i'd be trying to avoid (as to not expose sensitive data)

**Reply by DanRibbens** [reply]
@JPrisk,
Right, `id` will always be the PK. For UUIDs, whether they are generated from Payload or the database itself, would make non-guessable ids for you either way.

If your security is reliant on `id`s not being serial, your data can't be that sensitive or you need to come up with other measures besides security through obscurity.

**Reply by JPrisk** [reply]
That is very true, obscure `id`'s shouldn't be the only security measure for sensitive data. I can think of a few use cases where exposing the number of entries _could_ inadvertently tell a user about the volume of data in a system.

Regardless, like you mentioned there are ways around this currently - the aim was to raise for consideration, if you don't think it is roadmap worthy feel free to close it out

**Reply by JPrisk** [reply]
on second look I see what you mean, 🤦‍♂️  helps if you read the docs [https://payloadcms.com/docs/fields/overview#customizable-id]( https://payloadcms.com/docs/fields/overview#customizable-id)

> An explicit id field can be declared in the fields array to override this behavior

got it! closing this one out as a learning opportunity... 😂  thanks @DanRibbens
